server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -234,14 +266,28 @@ class AsyncAction {
             this.request = request;
             this.listener = listener;
             if (request.concreteNodes() == null) {
-                resolveRequest(request, clusterService.state());
-                assert request.concreteNodes() != null;
+                if (optimizedTransportActionNames.contains(request.getClass().getSimpleName())) {
+                    this.concreteNodes = resolveConcreteNodes(request, clusterService.state());
+                    assert request.concreteNodes() == null;
+                } else {
+                    resolveRequest(request, clusterService.state());
+                    assert request.concreteNodes() != null;
+                    this.concreteNodes = null;
+                }
+            } else {
+                this.concreteNodes = null;
+            }
+            if (request.concreteNodes() == null) {
+                assert concreteNodes != null;
+                this.responses = new AtomicReferenceArray<>(concreteNodes.length);
+            } else {
+                this.responses = new AtomicReferenceArray<>(request.concreteNodes().length);
             }
-            this.responses = new AtomicReferenceArray<>(request.concreteNodes().length);
         }
 
         void start() {
-            final DiscoveryNode[] nodes = request.concreteNodes();
+            logger.info(request.getClass().getSimpleName());
```

Comment:
This would lead to heavy logging. Please remove this.

Reply:
Sure Vikas, I intended to work on the logs for testing. Will remove them prior to raising the PR.